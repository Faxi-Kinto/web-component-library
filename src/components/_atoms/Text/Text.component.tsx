import React, { ReactNode } from 'react';
import * as Styled from './Text.styles';
import { defaultFontSize } from '@faxi/web-css-utilities/constants';

type TextProps = {
  children: React.ReactNode;
  className?: string;
  fontSize?: string;
  color?: string;
};

export const Body = (props: TextProps): JSX.Element => {
  const {
    children,
    className,
    color = 'black',
    fontSize = defaultFontSize,
  } = props;

  return (
    <Styled.Body className={className} color={color} fontSize={fontSize}>
      {children}
    </Styled.Body>
  );
};

interface HeadingTextProps extends TextProps {
  children: ReactNode;
  color?: string;
  level: '1' | '2' | '3' | '4' | '5' | '6';
}

export const Heading = (props: HeadingTextProps): JSX.Element => {
  const { children, className, color = 'black', level } = props;

  switch (level) {
    case '1':
      return (
        <Styled.Heading1 className={className} color={color}>
          {children}
        </Styled.Heading1>
      );
    case '2':
      return (
        <Styled.Heading2 className={className} color={color}>
          {children}
        </Styled.Heading2>
      );
    case '3':
      return (
        <Styled.Heading3 className={className} color={color}>
          {children}
        </Styled.Heading3>
      );
    case '4':
      return (
        <Styled.Heading4 className={className} color={color}>
          {children}
        </Styled.Heading4>
      );
    case '5':
      return (
        <Styled.Heading5 className={className} color={color}>
          {children}
        </Styled.Heading5>
      );
    case '6':
      return (
        <Styled.Heading6 className={className} color={color}>
          {children}
        </Styled.Heading6>
      );
    default:
      return (
        <Styled.Heading1 className={className} color={color}>
          {children}
        </Styled.Heading1>
      );
  }
};

interface ParagraphTextProps extends TextProps {
  children: ReactNode;
  color?: string;
  level: '1' | '2';
}

export const Paragraph = (props: ParagraphTextProps): JSX.Element => {
  const { children, className, color = 'black', level } = props;

  switch (level) {
    case '1':
      return (
        <Styled.Paragraph1 color={color} className={className}>
          {children}
        </Styled.Paragraph1>
      );
    case '2':
      return (
        <Styled.Paragraph2 color={color} className={className}>
          {children}
        </Styled.Paragraph2>
      );
    default:
      return (
        <Styled.Paragraph1 color={color} className={className}>
          {children}
        </Styled.Paragraph1>
      );
  }
};

export const Link = (
  props: React.HTMLProps<HTMLAnchorElement>
): JSX.Element => {
  const {
    children,
    className,
    color,
    download,
    href,
    media,
    rel,
    target,
    type,
  } = props;

  return (
    <Styled.Link
      className={className}
      color={color}
      download={download}
      href={href}
      media={media}
      rel={rel}
      target={target}
      type={type}
    >
      {children}
    </Styled.Link>
  );
};

export const Emphasized = (props: TextProps): JSX.Element => {
  const { children, className, color = 'black' } = props;
  return (
    <Styled.Emphasized color={color} className={className}>
      {children}
    </Styled.Emphasized>
  );
};

export default {
  Body,
  Heading,
  Paragraph,
  Link,
  Emphasized,
};
