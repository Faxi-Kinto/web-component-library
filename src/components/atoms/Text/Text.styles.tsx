import styled from 'styled-components';
import { pxToRem } from '@faxi/web-css-utilities';

const sharedTypographyStyles = `
  padding: 0;
  margin: 0;
`;

export const Body = styled.span`
  ${sharedTypographyStyles}
  ${(props: { fontSize: string; color: string }) => {
    return `
  font-size: ${pxToRem(props.fontSize)}};
  color: ${props.color};`;
  }};

  font-weight: 300;
  white-space: pre-wrap;
`;

export const Heading1 = styled.h1`
  ${sharedTypographyStyles}

  font-size: ${pxToRem('34px')};
  line-height: ${pxToRem('47px')};
  font-weight: bold;
  color: ${props => props.color};
`;

export const Heading2 = styled.h2`
  ${sharedTypographyStyles}

  font-size: ${pxToRem('30px')};
  line-height: ${pxToRem('42px')};
  font-weight: bold;
  color: ${props => props.color};
`;

export const Heading3 = styled.h3`
  ${sharedTypographyStyles}

  font-size: ${pxToRem('24px')};
  line-height: ${pxToRem('34px')};
  font-weight: 500;
  color: ${props => props.color};
`;

export const Heading4 = styled.h4`
  ${sharedTypographyStyles}

  font-size: ${pxToRem('18px')};
  line-height: ${pxToRem('25px')};
  font-weight: bold;
  color: ${props => props.color};
`;

export const Heading5 = styled.h5`
  ${sharedTypographyStyles}

  font-size: ${pxToRem('14px')};
  line-height: ${pxToRem('20px')};
  font-weight: 400;
  color: ${props => props.color};
`;

export const Heading6 = styled.h6`
  ${sharedTypographyStyles}

  font-size: ${pxToRem('12px')};
  line-height: ${pxToRem('17px')};
  font-weight: 400;
  color: ${props => props.color};
`;

export const Paragraph1 = styled.p`
  ${sharedTypographyStyles}

  font-size: ${pxToRem('18px')};
  line-height: ${pxToRem('25px')};
  font-weight: 400;
  color: ${props => props.color};
`;

export const Paragraph2 = styled.p`
  ${sharedTypographyStyles}

  font-size: ${pxToRem('16px')};
  line-height: ${pxToRem('23px')};
  font-weight: 400;
  color: ${props => props.color};
`;

export const Small = styled.small`
 ${sharedTypographyStyles}

  font-size: ${pxToRem('14px')};
  line-height: ${pxToRem('20px')};
  font-weight: 400;
  color: ${props => props.color};
`;

export const Link = styled.a`
  ${sharedTypographyStyles}
  color: ${props => props.color};
`;

export const Emphasized = styled.span`
  font-size: ${pxToRem('14px')};
  color: ${props => props.color};
  font-weight: bold;
`;
