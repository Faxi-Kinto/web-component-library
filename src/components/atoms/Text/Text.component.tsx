import React, { ReactNode } from 'react';
import * as Styled from './Text.styles';
import { defaultFontSize } from '@faxi/web-css-utilities/constants';

type TextProps = {
  children: React.ReactNode;
  className?: string;
  fontSize?: string;
  color?: string;
};

export const Body = (props: TextProps): JSX.Element => {
  const {
    children,
    className,
    fontSize = defaultFontSize,
    color = 'black',
  } = props;

  return (
    <Styled.Body className={className} fontSize={fontSize} color={color}>
      {children}
    </Styled.Body>
  );
};

interface HeadingTextProps extends TextProps {
  level: '1' | '2' | '3' | '4' | '5' | '6';
  children: ReactNode;
  color?: string;
}

export const Heading = (props: HeadingTextProps): JSX.Element => {
  const { children, level, className, color = 'black' } = props;

  switch (level) {
    case '1':
      return (
        <Styled.Heading1 color={color} className={className}>
          {children}
        </Styled.Heading1>
      );
    case '2':
      return (
        <Styled.Heading2 color={color} className={className}>
          {children}
        </Styled.Heading2>
      );
    case '3':
      return (
        <Styled.Heading3 color={color} className={className}>
          {children}
        </Styled.Heading3>
      );
    case '4':
      return (
        <Styled.Heading4 color={color} className={className}>
          {children}
        </Styled.Heading4>
      );
    case '5':
      return (
        <Styled.Heading5 color={color} className={className}>
          {children}
        </Styled.Heading5>
      );
    case '6':
      return (
        <Styled.Heading6 color={color} className={className}>
          {children}
        </Styled.Heading6>
      );
    default:
      return (
        <Styled.Heading1 color={color} className={className}>
          {children}
        </Styled.Heading1>
      );
  }
};

interface ParagraphTextProps extends TextProps {
  level: '1' | '2';
  children: ReactNode;
  color?: string;
}

export const Paragraph = (props: ParagraphTextProps): JSX.Element => {
  const { children, level, color = 'black' } = props;

  switch (level) {
    case '1':
      return <Styled.Paragraph1 color={color}>{children}</Styled.Paragraph1>;
    case '2':
      return <Styled.Paragraph2 color={color}>{children}</Styled.Paragraph2>;
    default:
      return <Styled.Paragraph1 color={color}>{children}</Styled.Paragraph1>;
  }
};

export const Link = (
  props: React.HTMLProps<HTMLAnchorElement>
): JSX.Element => {
  const { children, href, target, download, media, rel, type, color } = props;

  return (
    <Styled.Link
      href={href}
      target={target}
      download={download}
      media={media}
      rel={rel}
      type={type}
      color={color}
    >
      {children}
    </Styled.Link>
  );
};

export const Emphasized = (props: TextProps): JSX.Element => {
  const { children, color = 'black' } = props;
  return <Styled.Emphasized color={color}>{children}</Styled.Emphasized>;
};

export default {
  Body,
  Heading,
  Paragraph,
  Link,
  Emphasized,
};
