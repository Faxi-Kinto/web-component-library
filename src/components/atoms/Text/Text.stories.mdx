import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';

import Text from './';

<Meta title="Core|Atoms/Text" />

# Text Component

This componet is responsible for all app typography. Its broken down into subcomponents
and is used as per the examples below.

## Body Text

This subcomponent is responsible for rendering body text.

#### Implementation

<Preview>
  <Story name="body text">
    <Text.Body>This is some body text.</Text.Body>
  </Story>
</Preview>

#### Props

<Props of={Text.Body} />

## Headings

This subcomponent is responsible for endering all headings

### Implementations

#### Heading level 1

<Preview>
  <Story name="level 1 heading">
    <Text.Heading level="1">This is a level 1 heading</Text.Heading>
  </Story>
</Preview>

#### Heading level 2

<Preview>
  <Story name="level 2 heading">
    <Text.Heading level="2">This is a level 2 heading</Text.Heading>
  </Story>
</Preview>

#### Heading level 3

<Preview>
  <Story name="level 3 heading">
    <Text.Heading level="3">This is a level 3 heading</Text.Heading>
  </Story>
</Preview>

#### Heading level 4

<Preview>
  <Story name="level 4 heading">
    <Text.Heading level="4">This is a level 4 heading</Text.Heading>
  </Story>
</Preview>

#### Heading level 5

<Preview>
  <Story name="level 5 heading">
    <Text.Heading level="5">This is a level 5 heading</Text.Heading>
  </Story>
</Preview>

#### Heading level 6

<Preview>
  <Story name="level 6 heading">
    <Text.Heading level="6">This is a level 6 heading</Text.Heading>
  </Story>
</Preview>

#### Props

<Props of={Text.Heading} />

## Paragraph

#### Implementations

<Preview>
  <Story name="paragraph 1">
    <Text.Paragraph level="1">This is Paragraph 1</Text.Paragraph>
  </Story>
  <Story name="paragraph 2">
    <Text.Paragraph level="2">This is Paragraph 2</Text.Paragraph>
  </Story>
</Preview>

## Link

This subcomponent is responsible for rendering links.

#### Implementations

<Preview>
  <Story name="link">
    <Text.Link href="/">This is a simple link</Text.Link>
  </Story>
  <Story name="inline link">
    <Text.Body>
      This is some body text with&nbsp;
      <Text.Link href="https://www.google.com" target="_blank">
        External Link
      </Text.Link>&nbsp;that opens in new tab.
    </Text.Body>
  </Story>
</Preview>

#### Props

<Props of={Text.Link} />
