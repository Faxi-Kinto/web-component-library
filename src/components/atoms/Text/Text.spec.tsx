import React from 'react';
import renderer from 'react-test-renderer';
import Text from './Text.component';

test('Text Body renders correctly', () => {
  const component = renderer.create(
    <Text.Body>This is some body text.</Text.Body>
  );

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('Text Heading with level 1 renders correctly', () => {
  const component = renderer.create(
    <Text.Heading level="1">This is a heading</Text.Heading>
  );

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('Text Heading with level 2 renders correctly', () => {
  const component = renderer.create(
    <Text.Heading level="2">This is a heading</Text.Heading>
  );

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('Text Heading with level 3 renders correctly', () => {
  const component = renderer.create(
    <Text.Heading level="3">This is a heading</Text.Heading>
  );

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('Text Link renders correctly', () => {
  const component = renderer.create(<Text.Link>This is a link</Text.Link>);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
