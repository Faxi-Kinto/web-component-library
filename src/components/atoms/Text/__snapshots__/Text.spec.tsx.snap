// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Body renders correctly 1`] = `
.c0 {
  padding: 0;
  margin: 0;
  font-size: 1rem;
  color: black;
  font-weight: 300;
  white-space: pre-wrap;
}

<span
  className="c0"
  color="black"
  fontSize="16px"
>
  This is some body text.
</span>
`;

exports[`Text Heading with level 1 renders correctly 1`] = `
.c0 {
  padding: 0;
  margin: 0;
  font-size: 2.125rem;
  line-height: 2.9375rem;
  font-weight: bold;
  color: black;
}

<h1
  className="c0"
  color="black"
>
  This is a heading
</h1>
`;

exports[`Text Heading with level 2 renders correctly 1`] = `
.c0 {
  padding: 0;
  margin: 0;
  font-size: 1.875rem;
  line-height: 2.625rem;
  font-weight: bold;
  color: black;
}

<h2
  className="c0"
  color="black"
>
  This is a heading
</h2>
`;

exports[`Text Heading with level 3 renders correctly 1`] = `
.c0 {
  padding: 0;
  margin: 0;
  font-size: 1.5rem;
  line-height: 2.125rem;
  font-weight: 500;
  color: black;
}

<h3
  className="c0"
  color="black"
>
  This is a heading
</h3>
`;

exports[`Text Link renders correctly 1`] = `
.c0 {
  padding: 0;
  margin: 0;
}

<a
  className="c0"
>
  This is a link
</a>
`;
